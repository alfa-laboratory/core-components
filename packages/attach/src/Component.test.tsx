import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Attach } from './index';

describe('Attach', () => {
    describe('Snapshots tests', () => {
        it('should match snapshot', () => {
            const { container } = render(<Attach />);

            expect(container).toMatchSnapshot();
        });
    });

    it('should forward ref to input', () => {
        const ref = jest.fn();
        const dataTestId = 'test-id';
        const { getByTestId } = render(<Attach ref={ref} dataTestId={dataTestId} />);

        expect(ref.mock.calls).toEqual([[getByTestId(dataTestId)]]);
    });

    it('should set `data-test-id` attribute to input', () => {
        const dataTestId = 'test-id';
        const { getByTestId } = render(<Attach dataTestId={dataTestId} />);

        expect(getByTestId(dataTestId).tagName).toBe('INPUT');
    });

    it('should set `noFileText`', () => {
        const noFileText = 'empty attach';
        const { container } = render(<Attach noFileText={noFileText} />);

        expect(container.textContent).toContain(noFileText);
    });

    describe('Classes tests', () => {
        it('should set `className` class', () => {
            const className = 'test-class';
            const { container } = render(<Attach className={className} />);

            expect(container.firstElementChild).toHaveClass(className);
        });

        it('should set `disabled` atribute', () => {
            const dataTestId = 'test-id';
            const { container, getByTestId } = render(
                <Attach disabled={true} dataTestId={dataTestId} />,
            );

            expect(container.firstElementChild).toHaveClass('disabled');
            expect(getByTestId(dataTestId)).toHaveAttribute('disabled');
        });
    });

    describe('Callbacks tests', () => {
        it('should call `onChange` prop', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Attach onChange={cb} dataTestId={dataTestId} />);

            const input = getByTestId(dataTestId) as HTMLInputElement;

            fireEvent.change(input, {
                target: {
                    files: [
                        {
                            name: 'test.txt',
                            type: 'application/text',
                        },
                    ],
                },
            });

            expect(cb).toBeCalledTimes(1);
            expect(input.files && input.files.length).toEqual(1);
        });

        it('should not call `onChange` prop if disabled', async () => {
            const cb = jest.fn();
            const dataTestId = 'test-id';
            const { getByTestId } = render(
                <Attach onChange={cb} dataTestId={dataTestId} disabled={true} />,
            );

            const input = getByTestId(dataTestId) as HTMLInputElement;

            await userEvent.upload(input, {
                name: 'test.txt',
                type: 'application/text',
            } as File);

            expect(cb).not.toBeCalled();
            expect(input.files && input.files.length).toEqual(0);
        });
    });

    describe('Files tests', () => {
        it('should render selected file name if one file selected', () => {
            const dataTestId = 'test-id';
            const { container, getByTestId } = render(<Attach dataTestId={dataTestId} />);

            const input = getByTestId(dataTestId) as HTMLInputElement;

            fireEvent.change(input, {
                target: {
                    files: [
                        {
                            name: 'test.txt',
                            type: 'application/text',
                        },
                    ],
                },
            });

            expect(container.textContent).toContain('test.txt');
        });

        it('should render selected files count if several files selected', () => {
            const dataTestId = 'test-id';
            const { container, getByTestId } = render(
                <Attach multiple={true} dataTestId={dataTestId} />,
            );

            const input = getByTestId(dataTestId) as HTMLInputElement;

            fireEvent.change(input, {
                target: {
                    files: [
                        {
                            name: 'test1.txt',
                            type: 'application/text',
                        },
                        {
                            name: 'test2.txt',
                            type: 'application/text',
                        },
                        {
                            name: 'test3.txt',
                            type: 'application/text',
                        },
                    ],
                },
            });

            expect(container.textContent).toContain('3 файла');
        });

        it('should render "no file" and clear input value after clear button was clicked', () => {
            const dataTestId = 'test-id';
            const { container, getByTestId } = render(<Attach dataTestId={dataTestId} />);

            const input = getByTestId(dataTestId) as HTMLInputElement;

            fireEvent.change(input, {
                target: {
                    files: [
                        {
                            name: 'test.txt',
                            type: 'application/text',
                        },
                    ],
                },
            });

            expect(container.textContent).toContain('test.txt');

            const clearButton = container.querySelector('.clear');

            if (clearButton) {
                fireEvent.click(clearButton);
            }

            expect(container.textContent).toContain('Нет файла');
            expect(input.value).toBeFalsy();
        });

        it('should truncate the filename if the maxFilenameLength prop is passed', () => {
            const dataTestId = 'test-id';
            const { container, getByTestId } = render(
                <Attach maxFilenameLength={30} dataTestId={dataTestId} />,
            );

            const input = getByTestId(dataTestId) as HTMLInputElement;

            fireEvent.change(input, {
                target: {
                    files: [
                        {
                            name: 'so_long_filename_it_definitely_has_more_than_30_symbols.txt',
                            type: 'application/text',
                        },
                    ],
                },
            });

            expect(container.textContent).toContain('so_long_filena…30_symbols.txt');

            const clearButton = container.querySelector('.clear');

            if (clearButton) {
                fireEvent.click(clearButton);
            }

            fireEvent.change(input, {
                target: {
                    files: [
                        {
                            name: 'it_has_just_26_symbols.txt',
                            type: 'application/text',
                        },
                    ],
                },
            });

            expect(container.textContent).toContain('it_has_just_26_symbols.txt');
        });
    });

    it('should unmount without errors', () => {
        const { unmount } = render(<Attach />);

        expect(unmount).not.toThrowError();
    });
});
