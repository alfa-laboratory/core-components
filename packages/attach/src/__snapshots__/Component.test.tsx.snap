// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attach Snapshots tests should match snapshot 1`] = `
<div>
  <div
    class="component s"
  >
    <button
      class="component outlined s"
    >
      <span
        class="addons"
      >
        <svg
          class="icon"
          focusable="false"
          height="18"
          viewBox="0 0 17 18"
          width="17"
        >
          <g
            fill="none"
            fill-rule="evenodd"
          >
            <path
              d="M-3-3h24v24H-3"
            />
            <path
              d="M8.508 16.07a5.002 5.002 0 0 0 7.026-.044A4.998 4.998 0 0 0 15.578 9L7.8 1.222a3.996 3.996 0 0 0-6.69 1.793 3.999 3.999 0 0 0 1.034 3.864l6.364 6.363a3 3 0 0 0 5.037-1.337 2.998 2.998 0 0 0-.795-2.907L9.92 6.172l-.706.707 2.83 2.827a2.002 2.002 0 0 1-2.831 2.829l-6.36-6.364a2.996 2.996 0 0 1-.001-4.241 2.997 2.997 0 0 1 4.242 0l7.778 7.777a4 4 0 0 1-5.657 5.658L2.85 9l-.707.707 6.365 6.364z"
              fill="#0B1F35"
              fill-rule="nonzero"
            />
          </g>
        </svg>
      </span>
      <span
        class="text"
      >
        <span>
          Выберите файл
        </span>
      </span>
    </button>
    <label
      class="label"
    >
      <input
        class="control"
        tabindex="-1"
        type="file"
      />
    </label>
    <div
      class="noFile"
    >
      Нет файла
    </div>
  </div>
</div>
`;
